### PR TITLE
Various additions and fixes required for shadow support

### LINKED ISSUE #47: getlogin() unimplemented
While testing the coreutils programs the following utilities required `getlogin()`

- `logname`

### PR DESCRIPTION
**Pull request by Dennisbonke**
This PR aims to start implementing the required functionality for `shadow`. Some major group and user handling functions are still missing.
So far, this PR includes the following meta changes:
- We now build a dummy `libcrypt` to satisfy shadow.

The following functions have been added:
- `putspent()`,
- `lckpwdf()`,
- `ulckpwdf()`,
- `putpwent()`,
- `fgetpwent()`,
- `endusershell()`,
- `getusershell()`,
- `setusershell()`,
- `pututline()`,
- `putgrent()`,
- `fgetgrent()`,

The following functions have been implemented:
- `alphasort()`, this is implemented like musl,
- `getpass()`, with the `GLIBC` extension to read from `stdin` and write to `stdout` if `/dev/tty` cannot be opened,
- `getlogin()`,
- `sysconf()` now knows a response for the `sysconf(_SC_NGROUPS_MAX)` call

The following structs and constants have been added:
- The `SHADOW` constant, an alias for `_PATH_SHADOW`,
- The `spwd` struct,
- The `TIOCNOTTY` constant,
- The `lastlog` struct,

Furthermore, some miscellaneous fixes and additions include:
- `strtol()` is now able to handle negative numbers in bases != 10,
- Several headers have been added,
- A bug was found and fixed in `getpwnam()`.

This PR fixes #47

### CODE REVIEW

**Comment by Dennisbonke** [comment]
Please hold off on merging for a small moment, I want to add a bugfix regarding `__dlapi_reverse()` to this pr

**Comment by Dennisbonke** [comment]
Alright, fix is pushed, this can be merged now

File: options/ansi/generic/stdlib-stubs.cpp [hunk 1]
```diff
@@ -87,8 +87,7 @@ long strtol(const char *__restrict string, char **__restrict end, int base) {
 		string++;
 	}
 
-	if(base != 10) { // This is an ugly hack!
-		__ensure(!negative);
+	if(base != 10) {
 		return strtoul(string, end, base);
 	}
```
Comment (avdgrinten) [inline options/ansi/generic/stdlib-stubs.cpp:92]: This does not look correct. Don't we "forget" about the negative sign?

File: options/glibc/generic/shadow-stubs.cpp [hunk 2]
```diff
@@ -0,0 +1,26 @@
+#include <shadow.h>
+
+#include <bits/ensure.h>
+#include <mlibc/debug.hpp>
+
+#define NUM(n) ((n) == -1 ? 0 : -1), ((n) == -1 ? 0 : (n))
+#define STR(s) ((s) ? (s) : "")
```
Comment (avdgrinten) [inline options/glibc/generic/shadow-stubs.cpp:7]: Use lambda functions (within putspent) instead of macros here.

**Review by avdgrinten — COMMENTED** [review]
Some comments below, and the attribution of code that originates from musl needs to be fixed.

File: options/posix/generic/unistd-stubs.cpp [hunk 3]
```diff
@@ -724,9 +728,49 @@ int getpagesize() {
 	return mlibc::page_size;
 }
 
-char *getpass(const char *) {
-	__ensure(!"Not implemented");
-	__builtin_unreachable();
+// Code taken from musl
+// GLIBC extension for stdin/stdout
+char *getpass(const char *prompt) {
+	mlibc::infoLogger() << "\e[31mmlibc: getpass is broken, tcdrain is unavailable!\e[39m" << frg::endlog;
```
Comment (avdgrinten) [inline options/posix/generic/unistd-stubs.cpp:734]: It is probably a better idea to add a log to tcdrain() and let tcdrain() fail with ENOSYS. This function does not care about tcdrain()'s result anyway.

File: options/ansi/generic/stdlib-stubs.cpp [hunk 4]
```diff
@@ -87,9 +87,11 @@ long strtol(const char *__restrict string, char **__restrict end, int base) {
...
 
-	if(base != 10) { // This is an ugly hack!
-		__ensure(!negative);
-		return strtoul(string, end, base);
+	if(base != 10) {
+		unsigned long number = strtoul(string, end, base);
+		if(negative)
+			return static_cast<long>((~number) + 1);
+		return number;
```
Comment (avdgrinten) [inline options/ansi/generic/stdlib-stubs.cpp:94]: We can do this for all bases.

**Review by avdgrinten — COMMENTED** [review]
Some comments; the PR can be merged when they are fixed.

**Review by avdgrinten — APPROVED** [review]
LGTM.

**Pull request merged**